src/io/free_functions.rs

```diff
@@ -0,0 +1,293 @@
+        "jpg" | "jpeg" => jpeg::JPEGEncoder::new(fout).encode(buf, width, height, color),
+        #[cfg(feature = "png_codec")]
+        "png" => png::PNGEncoder::new(fout).encode(buf, width, height, color),
+        #[cfg(feature = "pnm")]
+        "pbm" => pnm::PNMEncoder::new(fout)
+            .with_subtype(pnm::PNMSubtype::Bitmap(pnm::SampleEncoding::Binary))
+            .encode(buf, width, height, color),
+        #[cfg(feature = "pnm")]
+        "pgm" => pnm::PNMEncoder::new(fout)
+            .with_subtype(pnm::PNMSubtype::Graymap(pnm::SampleEncoding::Binary))
+            .encode(buf, width, height, color),
+        #[cfg(feature = "pnm")]
+        "ppm" => pnm::PNMEncoder::new(fout)
+            .with_subtype(pnm::PNMSubtype::Pixmap(pnm::SampleEncoding::Binary))
+            .encode(buf, width, height, color),
+        #[cfg(feature = "pnm")]
+        "pam" => pnm::PNMEncoder::new(fout).encode(buf, width, height, color),
+        #[cfg(feature = "bmp")]
+        "bmp" => bmp::BMPEncoder::new(fout).encode(buf, width, height, color),
+        #[cfg(feature = "tiff")]
+        "tif" | "tiff" => tiff::TiffEncoder::new(fout).encode(buf, width, height, color)
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, Box::new(e))), // FIXME: see https://github.com/image-rs/image/issues/921
+        format => Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            &format!("Unsupported image format image/{:?}", format)[..],
+        )),
+    }
+}
+
+pub(crate) fn save_buffer_with_format_impl(
+    path: &Path,
+    buf: &[u8],
+    width: u32,
+    height: u32,
+    color: color::ColorType,
+    format: ImageFormat,
+) -> io::Result<()> {
+    let fout = &mut BufWriter::new(File::create(path)?);
+
+    match format {
+        #[cfg(feature = "ico")]
+        image::ImageFormat::ICO => ico::ICOEncoder::new(fout).encode(buf, width, height, color),
+        #[cfg(feature = "jpeg")]
+        image::ImageFormat::JPEG => jpeg::JPEGEncoder::new(fout).encode(buf, width, height, color),
+        #[cfg(feature = "png_codec")]
+        image::ImageFormat::PNG => png::PNGEncoder::new(fout).encode(buf, width, height, color),
+        #[cfg(feature = "bmp")]
+        image::ImageFormat::BMP => bmp::BMPEncoder::new(fout).encode(buf, width, height, color),
+        #[cfg(feature = "tiff")]
+        image::ImageFormat::TIFF => tiff::TiffEncoder::new(fout)
+            .encode(buf, width, height, color)
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, Box::new(e))),
+        _ => Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            &format!("Unsupported image format image/{:?}", format)[..],
+        )),
+    }
+}
+
+/// Guess format from a path.
+/// If the path has no extension or it can not be convert to a `str` for comparison then this
+/// method will simply return `None`.
+pub(crate) fn guess_format_from_path_impl(path: &Path) -> Result<ImageFormat, PathError> {
```

Comment:
This comment is no longer true

Reply:
This comment is no longer true :grinning: